### PR TITLE
Disable SDPA FlashAttention backward and mem eff attention on sm86+ for head_dim above 64

### PR DESCRIPTION
**Pull request by cpuhrsch**
Expand sdpa_utils.h check to disable FlashAttention when using autograd and mem eff attention for the following cases
- head_dim > 64
- sm86 or newer

Previously we only disable these kernels on sm86 and for head_dim equal to 128.

### CODE REVIEW

File: aten/src/ATen/native/transformers/cuda/sdp_utils.h [hunk 1]
```diff
@@ -533,12 +533,22 @@ inline bool check_gpu_sm86_head_dim_128(sdp_params params, bool debug) {
...
+    bool debug) {
+  // Flash Attention will raise an error in the backward pass if the head_dim
+  // size is 128 And the device is not sm80, the other head_dim check catches
+  // everything but sm86
+  auto dprops = at::cuda::getCurrentDeviceProperties();
+  bool is_sm86_or_newer = (dprops->major == 8) && (dprops->minor >= 6);
+  bool is_sm90_or_newer = (dprops->major == 9) && (dprops->minor >= 0);
+  if (!check_requires_grad(params, false) &&
+      !check_gpu_sm86_head_dim_128(params, false) &&
+      !is_sm86_or_newer &&
```
Comment (ngimel) [inline aten/src/ATen/native/transformers/cuda/sdp_utils.h:547]: I left a comment on another version of this PR, but `is_sm86_or_newer` will evaluate to `true` and this line will evaluate to `false`

File: test/test_transformers.py [hunk 2]
```diff
@@ -1412,7 +1412,7 @@ def test_flash_backward_sm86_headdim128(self):
         device = 'cuda'
         dtype = torch.float16
         make_tensor = partial(self.rand_tensor, type="dense", device=device, dtype=dtype)
-        # See check_gpu_sm86_head_dim_128 in pytorch/aten/src/ATen/native/transformers/cuda/sdp_utils.h
+        # See check_requires_grad_and_head_dim_gt64_and_sm_gt80 in pytorch/aten/src/ATen/native/transformers/cuda/sdp_utils.h
```
Comment (cpuhrsch) [inline test/test_transformers.py:1415]: I know, it's `check_requires_grad_and_head_dim_gt64_and_sm_gt86` not `check_requires_grad_and_head_dim_gt64_and_sm_gt80`

File: test/test_transformers.py [hunk 3]
```diff
@@ -57,6 +57,7 @@ def use_deterministic_algorithims(mode: bool, warn_only: bool):
 default_rtol = {torch.float16: 1e-3, torch.bfloat16: 1.6e-2, torch.float32: 1.3e-6}
 
 isSM86Device = torch.cuda.is_available() and torch.cuda.get_device_capability() == (8, 6)
+isSM89Device = torch.cuda.is_available() and torch.cuda.get_device_capability() == (8, 9)
```
Comment (malfet) [inline test/test_transformers.py:60]: Nit
```suggestion
isSM86or89Device = torch.cuda.is_available() and torch.cuda.get_device_capability() in [(8, 6), (8, 9])
```

File: test/test_transformers.py [hunk 4]
```diff
@@ -1395,29 +1396,39 @@ def test_sdp_choice_with_determinism(self, warn_only):
                 assert torch._fused_sdp_choice(query, key, value) == (
                     SDPBackend.EFFICIENT_ATTENTION if warn_only else SDPBackend.MATH)
 
-    @unittest.skipIf(not PLATFORM_SUPPORTS_FUSED_SDPA or not isSM86Device, "Does not support fused SDPA or not SM86 hardware")
-    def test_memory_efficeint_sm86_failure(self):
+    @unittest.skipIf(not PLATFORM_SUPPORTS_FUSED_SDPA or not (isSM86Device or isSM89Device),
```
Comment (malfet) [inline test/test_transformers.py:1399]:
```suggestion
    @unittest.skipIf(not PLATFORM_SUPPORTS_FUSED_SDPA or not isSM86or89Device,
```

File: test/test_transformers.py [hunk 5]
```diff
@@ -1395,29 +1396,39 @@ def test_sdp_choice_with_determinism(self, warn_only):
...
+        # See check_head_dim_gt64_and_sm_ge86 in pytorch/aten/src/ATen/native/transformers/cuda/sdp_utils.h
+        size = (2, 2, 4, head_dim)
         q, k, v = make_tensor(size), make_tensor(size), make_tensor(size)
         with sdp_kernel(enable_mem_efficient=True, enable_flash=False, enable_math=False):
             self.assertRaises(RuntimeError, lambda: torch.nn.functional.scaled_dot_product_attention(
                 q, k, v, None, 0.0, False))
 
-    @unittest.skipIf(not PLATFORM_SUPPORTS_FUSED_SDPA or not isSM86Device, "Does not support fused SDPA or not SM86 hardware")
-    def test_flash_backward_sm86_headdim128(self):
+    @unittest.skipIf(not PLATFORM_SUPPORTS_FUSED_SDPA or not (isSM86Device or isSM89Device),
```
Comment (malfet) [inline test/test_transformers.py:1413]:
```suggestion
    @unittest.skipIf(not PLATFORM_SUPPORTS_FUSED_SDPA or not isSM86or89Device,
```

File: test/test_transformers.py [hunk 6]
```diff
@@ -1627,8 +1638,8 @@ def test_mem_efficient_attention_vs_math_ref_grads(self, batch_size: int, seq_le
 
         # Create real output
         with sdp_kernel(enable_mem_efficient=True, enable_flash=False, enable_math=False):
-            # See check_gpu_sm86_head_dim_128 in pytorch/aten/src/ATen/native/transformers/cuda/sdp_utils.h
-            if isSM86Device and head_dim == 128:
+            # See check_head_dim_gt64_and_sm_ge86 in pytorch/aten/src/ATen/native/transformers/cuda/sdp_utils.h
+            if (isSM86Device or isSM89Device) and (head_dim > 64 and head_dim <= 128):
```
Comment (malfet) [inline test/test_transformers.py:1642]: I know range(64, 129) is a huge overkill there, but it looks syntactically pleasing, isn't it?
```suggestion
            if isSM86or89Device and head_dim in range(65, 129):
```

**Review by malfet — APPROVED** [review]
LGTM, but please consider using `isSM86or89Device` instead of pair of defines

**Pull request closed without merging**